core/drivers/crypto/caam/hal/common/hal_jr.c

```diff
@@ -0,0 +1,319 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   CAAM Job Rings Hardware Abstration Layer.
+ *          Implementation of primitives to access HW
+ */
+#include <caam_common.h>
+#include <caam_hal_jr.h>
+#include <caam_io.h>
+#include <caam_pwr.h>
+#include <caam_utils_delay.h>
+#include <registers/ctrl_regs.h>
+#include <registers/jr_regs.h>
+
+/*
+ * List of common JR registers to save/restore
+ */
+const struct reglist jr_backup[] = {
+	{ JRx_IRBAR, 2, 0, 0 },
+	{ JRx_IRSR, 1, 0, 0 },
+	{ JRx_ORBAR, 2, 0, 0 },
+	{ JRx_ORSR, 1, 0, 0 },
+	{ JRx_JRCFGR_LS, 1, 0, 0 },
+};
+
+/*
+ * Resets the Job Ring to ensure that all pending job completed
+ * and no other will be executed
+ *
+ * @baseaddr   Job Ring Base address
+ */
+enum CAAM_Status caam_hal_jr_reset(vaddr_t baseaddr)
+{
+	uint16_t timeout = 10000;
+	uint32_t reg_val = 0;
+
+	/*
+	 * Reset is done in 2 steps:
+	 *  - Flush all pending jobs (Set RESET bit)
+	 *  - Reset the Job Ring (Set RESET bit second time)
+	 */
+
+	/* Mask interrupts to poll for reset completion status */
+	io_mask32(baseaddr + JRx_JRCFGR_LS, JRx_JRCFGR_LS_IMSK,
+		  JRx_JRCFGR_LS_IMSK);
+
+	/* Initiate flush (required prior to reset) */
+	io_caam_write32(baseaddr + JRx_JRCR, JRx_JRCR_RESET);
+
+	do {
+		caam_udelay(100);
+		reg_val = io_caam_read32(baseaddr + JRx_JRINTR);
+		reg_val &= BM_JRx_JRINTR_HALT;
+	} while ((reg_val == JRINTR_HALT_ONGOING) && --timeout);
+
+	if (!timeout || reg_val != JRINTR_HALT_DONE) {
+		EMSG("Failed to flush job ring\n");
+		return CAAM_FAILURE;
+	}
+
+	/* Initiate reset */
+	timeout = 100;
+	io_caam_write32(baseaddr + JRx_JRCR, JRx_JRCR_RESET);
+	do {
+		caam_udelay(100);
+		reg_val = io_caam_read32(baseaddr + JRx_JRCR);
+	} while ((reg_val & JRx_JRCR_RESET) && --timeout);
+
+	if (!timeout) {
+		EMSG("Failed to reset job ring\n");
+		return CAAM_FAILURE;
+	}
+
+	return CAAM_NO_ERROR;
+}
+
+/*
+ * Configures the Job Ring HW queues.
+ * Returns the number of jobs complete.
+ *
+ * @baseaddr   Job Ring Base Address
+ * @nbjobs     Number of job rings supported
+ * @inrings    physical address of the JR input queue
+ * @outrings   physical address of the JR output queue
+ */
+void caam_hal_jr_config(vaddr_t baseaddr, uint8_t nbjobs, uint64_t inrings,
+			uint64_t outrings)
+{
+	uint32_t value = 0;
+
+	/* Setup the JR input queue */
+	io_caam_write32(baseaddr + JRx_IRBAR, ((inrings >> 32) & 0xFFFFFFFF));
+	io_caam_write32(((baseaddr + JRx_IRBAR) + 4), (inrings & 0xFFFFFFFF));
```

Comment:
The & 0xffffffffff here is redundant, `io_caam_write32()` if properly implemented should not be affected by the upper bits.